### ISSUE TITLE
[BUG] Infinite Loop in BlobContainerClient.listBlobs

### ISSUE THREAD

**Issue by atris** [Azure/azure-sdk-for-java#20515]
**Describe the bug**
Using forEach on listBlobs breaks since it gets into an infinite loop

***Exception or Stack Trace***
Add the exception log and stack trace if available

**To Reproduce**
```
        Collection<InetSocketAddress> addrs = new ArrayList<>();

        Iterator<BlobItem> blobItemIterator = blobContainerClient.listBlobs().iterator();

        while (blobItemIterator.hasNext()) {
            BlobItem blobItem = blobItemIterator.next();

            try {
                if (!blobItem.isDeleted())
                    addrs.add(addrFromString(blobItem.getName()));
            }
            catch (Exception e) {
                throw new IgniteSpiException("Failed to get content from the container: " + containerName, e);
            }
        }
```

***Code Snippet***
POM:

```
-->
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <parent>
        <artifactId>ignite-parent</artifactId>
        <groupId>org.apache.ignite</groupId>
        <version>1</version>
        <relativePath>../../parent</relativePath>
    </parent>
    <modelVersion>4.0.0</modelVersion>

    <artifactId>ignite-azure</artifactId>
    <version>2.11.0-SNAPSHOT</version>
    <url>http://ignite.apache.org</url>

    <dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>com.azure</groupId>
                <artifactId>azure-sdk-bom</artifactId>
                <version>1.0.2</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
        </dependencies>
    </dependencyManagement>

    <dependencies>
        <dependency>
            <groupId>org.apache.ignite</groupId>
            <artifactId>ignite-core</artifactId>
            <version>${project.version}</version>
            <scope>compile</scope>
        </dependency>

        <dependency>
            <groupId>org.apache.ignite</groupId>
            <artifactId>ignite-core</artifactId>
            <version>${project.version}</version>
            <type>test-jar</type>
            <scope>test</scope>
        </dependency>

        <dependency>
            <groupId>org.apache.ignite</groupId>
            <artifactId>ignite-tools</artifactId>
            <version>${project.version}</version>
            <scope>test</scope>
        </dependency>

        <!-- https://mvnrepository.com/artifact/org.jetbrains/annotations -->
        <dependency>
            <groupId>org.jetbrains</groupId>
            <artifactId>annotations</artifactId>
            <version>16.0.3</version>
        </dependency>

        <dependency>
            <groupId>com.azure</groupId>
            <artifactId>azure-core-http-netty</artifactId>
            <version>1.0.0</version>
        </dependency>

        <dependency>
            <groupId>com.azure</groupId>
            <artifactId>azure-core</artifactId>
            <version>1.0.0</version>
        </dependency>

        <dependency>
            <groupId>com.azure</groupId>
            <artifactId>azure-storage-blob</artifactId>
            <version>12.0.0</version>
        </dependency>

        <dependency>
            <groupId>com.azure</groupId>
            <artifactId>azure-storage-common</artifactId>
            <version>12.0.0</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/javax.cache/cache-api -->
        <dependency>
            <groupId>javax.cache</groupId>
            <artifactId>cache-api</artifactId>
            <version>1.0.0</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/com.fasterxml.jackson.core/jackson-annotations -->
        <dependency>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-annotations</artifactId>
            <version>2.10.0</version>
        </dependency>

        <dependency>
            <groupId>com.azure</groupId>
            <artifactId>azure-security-keyvault-secrets</artifactId>
        </dependency>
        <dependency>
            <groupId>com.azure</groupId>
            <artifactId>azure-identity</artifactId>
        </dependency>

        <!-- https://mvnrepository.com/artifact/io.netty/netty-all -->
        <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-all</artifactId>
        </dependency>

        <dependency>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-databind</artifactId>
            <version>2.10.0</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/com.fasterxml.jackson.dataformat/jackson-dataformat-xml -->
        <dependency>
            <groupId>com.fasterxml.jackson.dataformat</groupId>
            <artifactId>jackson-dataformat-xml</artifactId>
            <version>2.10.0</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/com.fasterxml.jackson.datatype/jackson-datatype-jsr310 -->
        <dependency>
            <groupId>com.fasterxml.jackson.datatype</groupId>
            <artifactId>jackson-datatype-jsr310</artifactId>
            <version>2.10.0</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/com.fasterxml.jackson.module/jackson-module-jaxb-annotations -->
        <dependency>
            <groupId>com.fasterxml.jackson.module</groupId>
            <artifactId>jackson-module-jaxb-annotations</artifactId>
            <version>2.10.0</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/org.apache.logging.log4j/log4j-api -->
        <dependency>
            <groupId>org.apache.logging.log4j</groupId>
            <artifactId>log4j-api</artifactId>
            <version>2.14.1</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/org.apache.logging.log4j/log4j-core -->
        <dependency>
            <groupId>org.apache.logging.log4j</groupId>
            <artifactId>log4j-core</artifactId>
            <version>2.14.1</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/org.apache.logging.log4j/log4j-slf4j-impl -->
        <dependency>
            <groupId>org.apache.logging.log4j</groupId>
            <artifactId>log4j-slf4j-impl</artifactId>
            <version>2.14.1</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/io.netty/netty-buffer -->
        <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-buffer</artifactId>
            <version>4.1.54.Final</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/io.netty/netty-codec -->
        <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-codec</artifactId>
            <version>4.1.54.Final</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/io.netty/netty-codec-http2 -->
        <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-codec-http2</artifactId>
            <version>4.1.54.Final</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/io.netty/netty-codec-http -->
        <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-codec-http</artifactId>
            <version>4.1.54.Final</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/io.netty/netty-codec-socks -->
        <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-codec-socks</artifactId>
            <version>4.1.54.Final</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/io.netty/netty-common -->
        <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-common</artifactId>
            <version>4.1.54.Final</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/io.netty/netty-handler -->
        <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-handler</artifactId>
            <version>4.1.54.Final</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/io.netty/netty-handler-proxy -->
        <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-handler-proxy</artifactId>
            <version>4.1.54.Final</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/io.netty/netty-resolver -->
        <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-resolver</artifactId>
            <version>4.1.54.Final</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/io.netty/netty-transport -->
        <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-transport</artifactId>
            <version>4.1.54.Final</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/io.netty/netty-transport-native-epoll -->
        <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-transport-native-epoll</artifactId>
            <version>4.1.54.Final</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/io.netty/netty-transport-native-kqueue -->
        <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-transport-native-kqueue</artifactId>
            <version>4.1.54.Final</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/io.netty/netty-transport-native-kqueue -->
        <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-transport-native-kqueue</artifactId>
            <version>4.1.54.Final</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/io.netty/netty-transport-native-unix-common -->
        <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-transport-native-unix-common</artifactId>
            <version>4.1.54.Final</version>
        </dependency>

        <dependency>
            <groupId>io.projectreactor.netty</groupId>
            <artifactId>reactor-netty</artifactId>
        </dependency>
        <dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-api</artifactId>
            <version>1.7.7</version>
        </dependency>

        <dependency>
            <groupId>org.reactivestreams</groupId>
            <artifactId>reactive-streams</artifactId>
            <version>1.0.3</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/io.projectreactor/reactor-core -->
        <dependency>
            <groupId>io.projectreactor</groupId>
            <artifactId>reactor-core</artifactId>
            <version>3.3.0.RELEASE</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/io.projectreactor.netty/reactor-netty -->
        <dependency>
            <groupId>io.projectreactor.netty</groupId>
            <artifactId>reactor-netty</artifactId>
            <version>0.9.0.RELEASE</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/org.reflections/reflections -->
        <dependency>
            <groupId>org.reflections</groupId>
            <artifactId>reflections</artifactId>
            <version>0.9.12</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/org.slf4j/slf4j-simple -->
        <dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-simple</artifactId>
            <version>1.8.0-beta4</version>
            <scope>test</scope>
        </dependency>

        <!-- https://mvnrepository.com/artifact/org.codehaus.woodstox/stax2-api -->
        <dependency>
            <groupId>org.codehaus.woodstox</groupId>
            <artifactId>stax2-api</artifactId>
            <version>4.2.1</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/com.fasterxml.jackson.core/jackson-core -->
        <dependency>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-core</artifactId>
            <version>2.10.0</version>
        </dependency>

    </dependencies>


</project>
```

**Comment by gapra-msft** [comment]
Hi @atris 
Thank you for reporting this issue. We will take a look at what may be going wrong and get back to you as soon as we can.

**Comment by gapra-msft** [comment]
Hi @atris 
I attempted to reproduce your issue with the exact pom you linked and unfortunately I was unable to reproduce the behavior you are seeing. 

```
        BlobServiceClient serviceClient = new BlobServiceClientBuilder()
                .connectionString(connectionString)
                .httpClient(new NettyAsyncHttpClientBuilder().proxy(new ProxyOptions(ProxyOptions.Type.HTTP, new InetSocketAddress("localhost", 8888))).wiretap(true).build())
                .buildClient();

        BlobContainerClient containerClient = serviceClient.createBlobContainer(container);

        containerClient.getBlobClient("a").getAppendBlobClient().create();
        containerClient.getBlobClient("a/b").getAppendBlobClient().create();
        containerClient.getBlobClient("b").getAppendBlobClient().create();
        containerClient.getBlobClient("c").getAppendBlobClient().create();
        containerClient.getBlobClient("b/d").getAppendBlobClient().create();
        containerClient.getBlobClient("b/e").getAppendBlobClient().create();


        Collection<String> addrs = new ArrayList<>();

        Iterator<BlobItem> blobItemIterator = containerClient.listBlobs().iterator();

        while (blobItemIterator.hasNext()) {
            BlobItem blobItem = blobItemIterator.next();
            System.out.println(blobItem.getName());

            if (!blobItem.isDeleted())
                addrs.add(blobItem.getName());

        }
        containerClient.delete();
```

With the pom you provided, I was able to run this code and get the following output 

```
a
a/b
b
b/d
b/e
c
```

Would it be possible for you to create a simple repro app on your end that consistently reproduces the issue?

**Comment by ashafique2** [comment]
@atris , did you find any solution for this?
I am facing the same issue.
```
 blobs = blobContainerClient.listBlobs.().iterator();
while(blobs.hasNext()){
System.out.println(blobs.next().getName())
}
```
printing an infinite list as follows:
`a->b->c->a->b->c->`a and so on...

**Comment by atris** [comment]
> @atris , did you find any solution for this?
> I am facing the same issue.
> 
> ```
> blobs = blobContainerClient.listBlobs.().iterator();
> while(blobs.hasNext()){
> System.out.println(blobs.next().getName())
> }
> ```
> 
> printing an infinite list as follows:
> `a->b->c->a->b->c->`a and so on...

I am using a set to keep the names seen and break when I see a repeat. This is because the names start repeating only after all names have been published.

**Comment by gapra-msft** [comment]
As I mentioned above, we were unable to reproduce the issue with the sample code I pasted above and your exact pom.

If you can provide a simple repro app that reproduces your issue with the exact hierarchy of blobs, we can continue further investigation.

**Comment by catharpin** [comment]
For those hunting for this later (I was chasing it today) - this appears to have been fixed between 12.10.0 and 12.11.0.

Current (12.13.0 as of this writing) does not exhibit the bug.
